zerver/lib/message.py

```diff
@@ -554,3 +558,130 @@ def apply_unread_message_event(state, message):
 
     state[unread_key].append(new_obj)
     state[unread_key].sort(key=key_func)
+
+def find_and_store_to_insert_stream_msgs(user_profile,
+                                         all_stream_messages,
+                                         all_stream_subscription_logs,
+                                         all_messages_to_insert):
+    # type: (UserProfile, DefaultDict[int, List[Message]], DefaultDict[int, List[RealmAuditLog]], List[UserMessage]) -> None
+    def store_user_message_to_insert(message):
+        # type: (Message) -> None
+        message = UserMessage(user_profile=user_profile,
+                              message_id=message['id'], flags=0)
+        all_messages_to_insert.append(message)
+
+    for (stream_id, stream_messages) in all_stream_messages.items():
+        stream_subscription_logs = all_stream_subscription_logs[stream_id]
+
+        for log_entry in stream_subscription_logs:
+            if len(stream_messages) == 0:
+                continue
+            if log_entry.event_type == 'subscription_deactivated':
+                for stream_message in stream_messages:
+                    if stream_message['id'] <= log_entry.event_last_message_id:
+                        store_user_message_to_insert(stream_message)
+                    else:
+                        break
+            elif log_entry.event_type in ('subscription_activated',
+                                          'subscription_created'):
+                initial_msg_count = len(stream_messages)
+                for i, stream_message in enumerate(stream_messages):
+                    if stream_message['id'] > log_entry.event_last_message_id:
+                        stream_messages = stream_messages[i:]
+                        break
+                final_msg_count = len(stream_messages)
+                if initial_msg_count == final_msg_count:
+                    if stream_messages[-1]['id'] <= log_entry.event_last_message_id:
+                        stream_messages = []
+            else:
+                raise AssertionError('%s is not a Subscription Event.' % (log_entry.event_type))
+
+        if len(stream_messages) > 0:
+            # We do this check for last event since if the last subscription
+            # event was a subscription_deactivated then we don't want to create
+            # UserMessage rows for any of the remaining messages.
+            if stream_subscription_logs[-1].event_type in (
+                    'subscription_activated',
+                    'subscription_created'):
+                for stream_message in stream_messages:
+                    store_user_message_to_insert(stream_message)
+
+def add_missing_messages(user_profile):
+    # type: (UserProfile) -> None
+    # This list will store all the messages for which we eventually will create
+    # UserMessage table rows by doing a bulk insert.
+    all_messages_to_insert = []  # type: List[UserMessage]
```

Comment:
This `all_messages_to_insert` should just be a list returned by `find_and_store_to_insert_stream_msgs`; I went ahead and changed the API to do that.